jpa/subsystem/src/main/java/org/jboss/as/jpa/container/EntityManagerCacheGlobal.java

```diff
@@ -0,0 +1,95 @@
+ *  under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation; either version 2.1 of
+ *  the License, or (at your option) any later version.
+ *
+ *  This software is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this software; if not, write to the Free
+ *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ *  02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.jpa.container;
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.persistence.EntityManager;
+
+import org.jipijapa.plugin.spi.EntityManagerCache;
+
+/**
+ * EntityManagerCacheGlobal is a Global (per application deployment/per persistence unit cache.
+ * EntityManagerCacheGlobal is thread-safe.
+ *
+ * @author Scott Marlow
+ */
+public class EntityManagerCacheGlobal implements EntityManagerCache {
+
+    private final ConcurrentLinkedDeque<EntityManager> cache = new ConcurrentLinkedDeque<>();
+    private final AtomicInteger cache_size = new AtomicInteger();
+    private final int max_size;
+
+    /**
+     *
+     * @param max_size if greater than zero, will use bounded cache, otherwise cache can have unlimited # of EntityManagers.
+     */
+    public EntityManagerCacheGlobal(int max_size) {
+        this.max_size = max_size;
+    }
+
+    @Override
+    public EntityManager get() {
+        EntityManager result;
+        do {
+            if ( null == (result = cache.poll())) {
+                // return null if cache is empty
+                return null;
+            }
+            if (max_size > 0) {
+                cache_size.getAndDecrement();
+            }
+        } while(!result.isOpen());
+        return result;
+    }
+
+    @Override
+    public void put(EntityManager entityManager) {
+        boolean cacheHasRoom = max_size < 1 || cache_size.get() < max_size;
+        // put in cache if there is enough room for it
+        if (cacheHasRoom) {
+            if (max_size > 0) {
+                // check if previous cache_size value is still less than max_size
+                cacheHasRoom = (cache_size.getAndIncrement() < max_size);
+            }
+            if (cacheHasRoom) {
+                cache.push(entityManager);
+            }
+        }
+
+        // if cache doesn't have enough room, close the entity manager.
+        if(!cacheHasRoom) {
+            if (entityManager.isOpen()) {
+                entityManager.close();
+            }
+        }
+    }
+
+    // NOTE: only expected to be called at application undeployment time.
+    @Override
+    public void clear() {
+        cache.clear();
```

Comment:
Shouldn't this drain the queue and close any open entity managers?

Reply:
It should be faster to skip closing each open entity manager and instead close the underlying entity manager factory, which we are doing in org.jboss.as.jpa.service.PersistenceUnitServiceImpl.stop(StopContext).